File: app/src/main/java/org/michaelbel/moviemade/presentation/features/main/MoviesAdapter.java
package org.michaelbel.moviemade.presentation.features.main;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.Glide;

import org.michaelbel.moviemade.R;
import org.michaelbel.moviemade.core.entity.Movie;
import org.michaelbel.moviemade.core.utils.DeviceUtil;
import org.michaelbel.moviemade.core.utils.TmdbConfigKt;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.internal.DebouncingOnClickListener;

public class MoviesAdapter extends RecyclerView.Adapter<MoviesAdapter.MoviesViewHolder> {

    public interface Listener {
        void onMovieClick(Movie movie, View view);
    }

    private Listener movieClickListener;
    private List<Movie> movies = new ArrayList<>();

    public MoviesAdapter(Listener listener) {
        movieClickListener = listener;
    }

    public List<Movie> getMovies() {
        return movies;
    }

    public void addMovies(List<Movie> movies) {
        this.movies.addAll(movies);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MoviesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_poster, parent, false);
        MoviesViewHolder holder = new MoviesViewHolder(view);
        view.setOnClickListener(new DebouncingOnClickListener() {
            @Override
            public void doClick(View v) {
                int position = holder.getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    movieClickListener.onMovieClick(movies.get(position), v);
                }
            }
        });

        if (DeviceUtil.INSTANCE.isLandscape(parent.getContext()) || DeviceUtil.INSTANCE.isTablet(parent.getContext())) {
            view.getLayoutParams().height = (int) (parent.getWidth() / 2.5F);
        } else {
            view.getLayoutParams().height = (int) (parent.getWidth() / 2 * 1.5F);
        }

        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull MoviesViewHolder holder, int position) {
        Movie movie = movies.get(position);
        Glide.with(holder.itemView.getContext()).load(String.format(Locale.US, TmdbConfigKt.TMDB_IMAGE, "w342", movie.getPosterPath())).thumbnail(0.1f).into(holder.posterImage);
    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    class MoviesViewHolder extends RecyclerView.ViewHolder {

        AppCompatImageView posterImage;

        private MoviesViewHolder(View view) {
            super(view);
            posterImage = view.findViewById(R.id.poster);
        }
    }
}

File: app/src/main/java/org/michaelbel/moviemade/presentation/features/main/MoviesAdapter.kt
package org.michaelbel.moviemade.presentation.features.main

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.listitem_movie.*
import org.michaelbel.data.Movie
import org.michaelbel.moviemade.R
import org.michaelbel.moviemade.core.DeviceUtil
import org.michaelbel.moviemade.core.TmdbConfig.TMDB_IMAGE
import org.michaelbel.moviemade.presentation.common.DebouncingOnClickListener
import java.util.*

class MoviesAdapter(
        private val listener: Listener
): RecyclerView.Adapter<MoviesAdapter.MoviesViewHolder>() {

    interface Listener {
        fun onMovieClick(movie: Movie)
    }

    val movies = ArrayList<Movie>()

    fun addMovies(results: List<Movie>) {
        movies.addAll(results)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MoviesViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.listitem_movie, parent, false)
        val holder = MoviesViewHolder(view)

        val landscape = DeviceUtil.isLandscape(parent.context) || DeviceUtil.isTablet(parent.context)
        view.layoutParams.height = if (landscape) (parent.width / 2.5F).toInt() else (parent.width / 2 * 1.5F).toInt()

        return holder
    }

    override fun onBindViewHolder(holder: MoviesViewHolder, position: Int) {
        holder.bind(movies[position])
    }

    override fun getItemCount() = movies.size

    inner class MoviesViewHolder(override val containerView: View):
            RecyclerView.ViewHolder(containerView), LayoutContainer {

        fun bind(movie: Movie) {
            Glide.with(containerView.context)
                    .load(String.format(Locale.US, TMDB_IMAGE, "w342", movie.posterPath))
                    .thumbnail(0.1F)
                    .into(poster)

            containerView.setOnClickListener(object: DebouncingOnClickListener() {
                override fun doClick(v: View) {
                    if (adapterPosition != RecyclerView.NO_POSITION) {
                        listener.onMovieClick(movies[adapterPosition])
                    }
                }
            })
        }
    }
}